packages/metro/src/lib/contextModuleTemplates.js

```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ * @format
+ */
+
+import * as path from 'path';
+import type {ContextMode} from '../ModuleGraph/worker/collectDependencies';
+
+function createFileMap(
+  modulePath: string,
+  files: string[],
+  processModule: (moduleId: string) => string,
+): string {
+  let mapString = '\n';
+
+  files
+    .slice()
+    // Sort for deterministic output
+    .sort()
+    .forEach(file => {
+      let filePath = path.relative(modulePath, file);
+
+      // NOTE(EvanBacon): I'd prefer we prevent the ability for a module to require itself (`require.context('./')`)
+      // but Webpack allows this, keeping it here provides better parity between bundlers.
+
+      // Ensure relative file paths start with `./` so they match the
+      // patterns (filters) used to include them.
+      if (!filePath.startsWith('.')) {
+        filePath = `.${path.sep}` + filePath;
+      }
+      const key = JSON.stringify(filePath);
+      // NOTE(EvanBacon): Webpack uses `require.resolve` in order to load modules on demand,
+      // Metro doesn't have this functionality so it will use getters instead. Modules need to
+      // be loaded on demand because if we imported directly then users would get errors from importing
+      // a file without exports as soon as they create a new file and the context module is updated.
+
+      // NOTE: The values are set to `enumerable` so the `context.keys()` method works as expected.
+      mapString += `  ${key}: { enumerable: true, get() { return ${processModule(
+        file,
+      )}; } },\n`;
+    });
+  return `Object.defineProperties({}, {${mapString}})`;
+}
+
+function getEmptyContextModuleTemplate(modulePath: string, id: string): string {
+  return `
+function metroEmptyContext(request) {
+  let e = new Error("No modules for context '" + ${JSON.stringify(id)} + "'");
```

Comment:
@EvanBacon I'm making minor changes to this code and noticed an inconsistency - an empty context module will throw here, but trying to get a missing module from a non-empty context does not throw. Is this Webpack's behaviour as well?

Reply:
Ideally I would like to remove this ID from the body of the module entirely - otherwise we run into machine-independence issues (Windows/POSIX paths, paths needing to be relative to the project root...) which can cause all sorts of downstream problems with caching etc.

For this kind of diagnostic, we should instead pass the numeric `module.id` into a dedicated helper in `require.js` where we have access to module system metadata (such as paths) that module bodies cannot generally access.